packages/shared/src/events/eventing.ts

```diff
@@ -19,40 +19,62 @@ export type ErrorEvent = CommonEventProps;
 
-type EventMap = {
-  [AllProviderEvents.Ready]: ReadyEvent;
-  [AllProviderEvents.Error]: ErrorEvent;
-  [AllProviderEvents.Stale]: StaleEvent;
-  [AllProviderEvents.ContextChanged]: CommonEventProps;
-  [AllProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
+type ServerEventMap = {
+  [ServerProviderEvents.Ready]: ReadyEvent;
+  [ServerProviderEvents.Error]: ErrorEvent;
+  [ServerProviderEvents.Stale]: StaleEvent;
+  [ServerProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
 };
 
-export type EventContext<  U extends Record<string, unknown> = Record<string, unknown>
-> = EventMap[AllProviderEvents] & U;
+type ClientEventMap = {
+  [ClientProviderEvents.Ready]: ReadyEvent;
+  [ClientProviderEvents.Error]: ErrorEvent;
+  [ClientProviderEvents.Stale]: StaleEvent;
+  [ClientProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
+  [ClientProviderEvents.ContextChanged]: CommonEventProps;
+};
+
+export type NotServerChangeEvents = ServerProviderEvents.Ready | ServerProviderEvents.Error | ServerProviderEvents.Stale;
+export type NotClientChangeEvents = ClientProviderEvents.Ready | ClientProviderEvents.Error | ClientProviderEvents.Stale | ClientProviderEvents.ContextChanged;
+export type NotChangeEvents = NotServerChangeEvents | NotClientChangeEvents;
+
+export type EventContext<
+  U extends Record<string, unknown> = Record<string, unknown>,
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = (T extends ClientProviderEvents ? ClientEventMap[T] : T extends ServerProviderEvents ? ServerEventMap[T] : never) &
+  U;
 
-export type EventDetails = EventContext & CommonEventDetails;
-export type EventHandler = (eventDetails?: EventDetails) => Promise<unknown> | unknown;
+export type EventDetails<
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = EventContext<Record<string, unknown>, T> & CommonEventDetails;
+export type EventHandler<
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = (eventDetails?: EventDetails<T>) => Promise<unknown> | unknown;
 
-export interface Eventing {
+export interface Eventing<T extends ServerProviderEvents | ClientProviderEvents> {
   /**
    * Adds a handler for the given provider event type.
    * The handlers are called in the order they have been added.
-   * @param {AnyProviderEvent} eventType The provider event type to listen to
+   * @param eventType The provider event type to listen to
    * @param {EventHandler} handler The handler to run on occurrence of the event type
    */
-  addHandler(eventType: AnyProviderEvent, handler: EventHandler): void;
+  addHandler(
+    eventType: ClientProviderEvents.ConfigurationChanged | ServerProviderEvents.ConfigurationChanged,
+    handler: EventHandler<ClientProviderEvents.ConfigurationChanged | ServerProviderEvents.ConfigurationChanged>,
+  ): void;
```

Comment:
This means that eventType will never be more specific than  `ClientProviderEvents.ConfigurationChanged | ServerProviderEvents.ConfigurationChanged` which can be seen in the test: 

<img width="1517" alt="Bildschirmfoto 2024-02-14 um 21 31 49" src="https://github.com/open-feature/js-sdk/assets/68054864/a915d3b1-5c9a-4c3d-b8f2-adefcfbbbc7c">

Maybe we could go for a conditional type for the generic `T` of the eventing? It is a bit confusing to see the client types in the server.

This could look like this (I just had time to sketch this, we should maybe create explicit types for these inline conditionals): 
```typescript
export interface Eventing<T extends ServerProviderEvents | ClientProviderEvents> {
  /**
   * Adds a handler for the given provider event type.
   * The handlers are called in the order they have been added.
   * @param eventType The provider event type to listen to
   * @param {EventHandler} handler The handler to run on occurrence of the event type
   */
  addHandler(
    eventType: T extends ServerProviderEvents
      ? ServerProviderEvents.ConfigurationChanged
      : ClientProviderEvents.ConfigurationChanged,
    handler: EventHandler<
      T extends ServerProviderEvents
        ? ServerProviderEvents.ConfigurationChanged
        : ClientProviderEvents.ConfigurationChanged
    >,
  ): void;

  addHandler(
    eventType: T extends ServerProviderEvents ? NotServerChangeEvents : NotClientChangeEvents,
    handler: EventHandler<T extends ServerProviderEvents ? NotServerChangeEvents : NotClientChangeEvents>,
  ): void;

  /**
   * Removes a handler for the given provider event type.
   * @param {AnyProviderEvent} eventType The provider event type to remove the listener for
   * @param {EventHandler} handler The handler to remove for the provider event type
   */
  removeHandler(eventType: T, handler: EventHandler<T>): void;

  /**
   * Gets the current handlers for the given provider event type.
   * @param {AnyProviderEvent} eventType The provider event type to get the current handlers for
   * @returns {EventHandler[]} The handlers currently attached to the given provider event type
   */
  getHandlers(eventType: T): EventHandler<T>[];
}
```
<img width="1517" alt="Bildschirmfoto 2024-02-14 um 21 31 49" src="https://github.com/open-feature/js-sdk/assets/68054864/8ded3d8b-b2f0-42cf-92df-7581063e278a">

Reply:
Oh that makes sense. I like it. I will give that a shot, maybe with aliases as you say.

Reply:
I fixed this, and it doesn't even need to be generic here. I did so by putting a ternary like the one you show here in the `Eventing` interface itself, and then this implementation simply can restrict it further to `ProviderEvents` (which is the package-specific enum for server/client).

Reply:
@lukas-reining unfortunately GitHub is having [some issues](https://www.githubstatus.com/), so the latest changes aren't showing up on this PR yet, even though they are on the branch :exploding_head:

Reply:
It's working now.